wireguard: Add support for managed K8s

gandro opened the pull request:
This adds support for Wireguard on Cilium installations without a per-node pod CIDR, such as managed Kubernetes.
See the commit message for the implementation details.

I have tested this on AKS (with #15533 applied). Still need to find a away to experiment with EKS and GCP as they do not have recent enough Kernel versions, but none of the changes in this PR are cloud-vendor specific.

Ref #15462

edited by @aanm:
:warning: **note for reviewers**: The base branch of this PR is **not** the `master` branch,

gandro committed:
wireguard: Populate AllowedIPs from IPCache

Wireguard's Cryptokey Routing [1] requires each peer to specify the list
of source or destination IP addresses expected to be sent over the
tunnel. In Cilium, these IP addresses are the addresses of all pods (for
pod2pod traffic) hosted by the remote node.

Before this commit, Wireguard would use the remote nodes pod cidr as the
list of `AllowedIPs`. This approach however falls short on managed
Kubernetes installations, where Cilium allocates pod IPs via cloud
provider API.

We therefore need to populate the `AllowedIPs` with a complete list of
pod IPs we expect to observe for a given peer. This information is
available in Cilium's IPCache, which this commit now integrates into the
Wireguard setup code.

In order to be able to add and remove individual IPs from the
`AllowedIPs` list, we need to mirror the kernel state. This is
implemented by storing a `peerConfig` for each known Wireguard peer.
When a peer is first discovered via node manager, we obtain the remote
peers `AllowedIPs` by querying Cilium's user-space copy of the IPCache
in the agent. In addition, we also subscribe to IPCache updates in the
Wireguard agent and update the `AllowedIPs` list of known peers
accordingly.

Ref #15462

[1] https://www.wireguard.com/#cryptokey-routing

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

brb commented on pkg/wireguard/agent/agent.go:
```diff
@@ -364,3 +424,94 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
+
+// removedAllowedIP appends ip to the list of allowedIPs
+func (p *peerConfig) addAllowedIP(ip net.IPNet) {
+	p.allowedIPs = append(p.allowedIPs, ip)
```
Maybe a bit of a cargo cult programming, but shall we check for duplicates (and log a warning if such exist)? From reading the change, it seems possible to have the duplicates. E.g. fetching from the IPCache, and then receiving an event for an IP which we already fetched.

gandro replied:
I don't believe we should have duplicates. But might be worth adding a check to test this. Because we hold the IPCache mutex when fetching the initial set in `UpdatePeer` there should never be any queued/concurrent updates, i.e. we should (at least in theory) never receive an IP upsert event twice.

gandro replied:
Fixed. My assumption that we would never get an IP upsert twice was also wrong. It's actually very common that an identity changes. I changed the method to do a check.

brb commented on pkg/wireguard/agent/agent.go:
```diff
@@ -364,3 +424,94 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
+	updatedPeers := []*peerConfig{}
+
+	// Remove IP from allowedIPs if it was deleted or the host IP changed
+	if modType == ipcache.Delete || !newHostIP.Equal(oldHostIP) {
```
Shouldn't we handle the case when the host IP (aka node IP) changed only in `UpdatePeer()`? The latter will be invoked in any case upon the change.

If we could handle the change in `UpdatePeer()`, then both if's would become mutually exclusive, and we could get rid of `updatedPeers`.

gandro replied:
To be honest, I'm not sure what exactly happens when a node IP changes with regard to `UpdatePeer()`. I have to assume that a node IP change is not an atomic operation on the IPCache, i.e. we will get a bunch of `Upsert` events with `!newHostIP.Equal(oldHostIP)` here. I should probably test it (but I also understand that right now this is not yet fully implemented).

I think you actually convinced me that the current approach here is indeed not correct. Right now, we inherit the allowedIPs entries from the node before the IP change in `UpdatePeer()`. Which means we suddenly have a different view than the IPCache, which might still have the old `hostIP` for some entries. In that case, we'll miss events such as `Delete` events because they still have the old host IP set, while we already assume the new one. :thinking:

brb replied:
>  In that case, we'll miss events such as Delete events because they still have the old host IP set, while we already assume the new one.

Needs testing, but I *hope* that `UpdatePeer()` is called when the node IP addr changes. If it's guaranteed that `UpdatePeer()` happens before the IPCache updates, then we should easily handle this. But if there is no such guarantee, then :thinking:

gandro replied:
I have tested and verified that `UpdatePeer` is indeed invoked when the nodeIP changes. I have modified the code in a way that we always re-fetch the IPCache state whenever we detect a node IP change. So an IP change is basically identical to a node being added again. This way it does not matter if we see the `IPCache` events before or after we see the `UpdatePeer()` event. 

In other words, if there are IPCache events for an unknown node IP (new node or changed node) before we observe `UpdatePeer()` for that IP, the IPCache events are ignored. Once `UpdatePeer()` is called, we stay in sync.

However, I'm pretty sure that Cilium currently does not fully support node IP changes atm. The IPCache entries were not updated (which is out of scope for this PR).

brb commented on pkg/wireguard/agent/agent.go:
```diff
@@ -364,3 +424,94 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
+	}
+
+	// Add IP to allowedIP if it was added or the host IP changed
+	if modType == ipcache.Upsert || newHostIP != nil {
```
The same question as with the host IP change above.

gandro replied:
Removed.

brb commented on pkg/wireguard/agent/agent.go:
```diff
@@ -364,3 +424,94 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
+
+	// Remove IP from allowedIPs if it was deleted or the host IP changed
+	if modType == ipcache.Delete || !newHostIP.Equal(oldHostIP) {
+		oldNodeName := a.nodeNameByNodeIP[oldHostIP.String()]
```
I think the node name can be empty if we haven't finished the restore. I guess it's fine?

gandro replied:
Fixed, we now check if the nodeName exists in the map. IPCache events during restore are ignored until UpdatePeer is called for a peer.

brb commented on pkg/wireguard/agent/agent.go:
```diff
@@ -364,3 +424,94 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
+
+// OnIPIdentityCacheChange implements ipcache.IPIdentityMappingListener
+func (a *Agent) OnIPIdentityCacheChange(modType ipcache.CacheModification, ipnet net.IPNet, oldHostIP, newHostIP net.IP,
+	oldID *identity.NumericIdentity, newID identity.NumericIdentity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata) {
```
Just an idea, but as we currently encrypt only pod2pod, we could avoid unnecessary updates for identities which we don't care about (i.e. non-endpoint identities).

gandro replied:
I'm not sure what you are proposing. We are actually are ignoring updates for e.g. non-endpoint identities in the function body here, as we only consider entries where the `hostIP` (`oldHostIP`/`newHostIP`) is not `nil`. Meaning all updates with `hostIP=nil` are ignored by the body of `OnIPIdentityCacheChange`.

The current IPCache interface does not have a function to only subscribe to events where `hostIP` is set.

gandro replied:
Actually, I realize now that I should have added a `{new,old}HostIP!=nil` check to the `Delete` and `Upsert` case. While it was my intent to ignore updates without the `hostIP`, the current code does not make that clear. It only works by accident right now for cases where the `hostIP` is nil.

gandro replied:
Fixed. The code now should be much more explicit in the kind of events it handles.

brb reviewed (commented):
LGTM :rocket: A few comments / questions.

I've tested for transient packet drops during cilium-agent restart / pub key changes - it works!

brb committed:
wireguard: Remove operator

This commit removes the Wireguard operator which previously was used to
allocate a Wireguard tunnel IP.

After taking a closer look into the Wireguard source code in the kernel
and realising that we don't see the tunnel IP on the wire, we came to
realisation that the IP is not needed at all.

For host2host encryption with Wireguard (TODO) we don't need to rely on
the tunnel IP (we will be able to use the nodeIP and Wireguard's fwmark
to avoid the loop).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
wireguard: Add NodeIP to AllowedIPs list

This commit adds a remote NodeIPv{4,6} to the AllowedIPs to the
corresponding peer entry.

This is required, as after we have removed the Wireguard tunnel IP,
requests forwarded by NodePort BPF will have the NodeIP instead of the
tunnel IP. So the remote peer will receive such requests on "cilium_wg0"
with srcIP = NodeIP. If the NodeIP is not in the AllowedIPs list, the
request will be dropped by the Wireguard device driver.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
wireguard: Completely disable rp_filter

Just having a default route is not enough to pass the rp_filter check in
the loose mode. So, disable it completely.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on pkg/ipcache/ipcache.go:
```diff
@@ -580,6 +580,23 @@ func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (ips []string)
+// LookupByHostRLocked returns the list of IPs returns the set of IPs
+// (endpoint or CIDR prefix) that have hostIP associated as the host of the
+// entry. Requires the caller to hold the RLock.
+func (ipc *IPCache) LookupByHostRLocked(hostIP net.IP) (cidrs []net.IPNet) {
```
We can cut the lookup by half if we support sending a v4 and a v6 hostIP addresses as an argument of this function.

aanm commented on pkg/wireguard/agent/agent.go:
```diff
@@ -314,3 +354,113 @@ func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
+
+// OnIPIdentityCacheChange implements ipcache.IPIdentityMappingListener
+func (a *Agent) OnIPIdentityCacheChange(modType ipcache.CacheModification, ipnet net.IPNet, oldHostIP, newHostIP net.IP,
+	oldID *identity.NumericIdentity, newID identity.NumericIdentity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata) {
```
nit: all arguments in this line are not used so it would probably make sense to remove their variables. (e.g. `oldID *identity.NumericIdentity` -> `*identity.NumericIdentity`

gandro replied:
Will replace with `_` (Go does always needs an identifier)

aanm commented on pkg/wireguard/agent/agent.go:
```diff
@@ -236,21 +244,35 @@ func (a *Agent) UpdatePeer(nodeName, pubKeyHex string,
 	}
```
line 240 should be `option.Config.EnableIPv4 && nodeIPv4 != nil`. Similar changes for line 242

aanm commented on pkg/wireguard/agent/agent.go:
```diff
@@ -57,8 +61,9 @@ type Agent struct {
 	privKey          wgtypes.Key
-	pubKeyByNodeName map[string]string // nodeName => pubKey
-	restoredPubKeys  map[string]struct{}
+	peerByNodeName   map[string]*peerConfig
```
Why don't we store the node IP address as a key of this map instead of having the `nodeNameByNodeIP` with redundant information?

gandro replied:
I considered it, but it's not better. A peer will have two IPs (v4+v6), but only ever a single name. This makes the name the better choice as a key.

I need to lookup by name for `UpdatePeer` (to deal with IP change) and `DeletePeer`. This means I either need `nodeNameByIP` (which I have right now) or `nodeIPsByNodeName`, but the latter proved annoying because I suddenly had to keep two entries per node in the peer map.

aanm commented on pkg/wireguard/agent/agent.go:
```diff
@@ -174,58 +182,58 @@ func (a *Agent) RestoreFinished() error {
+		}
+		if option.Config.EnableIPv6 && nodeIPv6 != nil {
+			allowedIPs = append(allowedIPs, ipcache.IPIdentityCache.LookupByHostRLocked(nodeIPv6)...)
+			allowedIPs = append(allowedIPs, *iputil.IPToPrefix(nodeIPv6))
```
Similarly to what is done on line 456, shouldn't we catch on duplicates here? Hmm maybe we won't have duplicates in the ipcache, so it should probably be fine.

gandro replied:
That IPCache only has a single entry per IP is my assumption, yes.

aanm commented on pkg/wireguard/agent/agent.go:
```diff
@@ -121,7 +127,7 @@ func (a *Agent) Init() error {
 		return err
```
Line 107: `sysctl.Disable(fmt.Sprintf("net.ipv4.conf.%s.rp_filter", types.IfaceName))`

gandro replied:
I will leave this unchanged to avoid conflicts with this line in the upstream branch (it's completely unrelated to this PR). @brb feel free to adapt that in your commit.

brb replied:
https://github.com/cilium/cilium/pull/15565/commits/6148cc24914a4edce990839f87d7611674ebde67 addressed this.

aanm reviewed (changes requested):
It would be great to have unit tests testing the concurrency of event handling for node (`[Update|Delete]Peer`) and pod events (`OnIPIdentityCacheChange`).

Also, this PR has the title: "add support for managed k8s", in all of these environments a kvstore is never used. Thus, we could safely index CiliumEndpoints in the k8s watchers stores by node IP addresses. This would allow us to remove the function on `LookupByHostRLocked` and perform the lookup directly in the up to date state from the CiliumEndpoints store.

aanm reviewed (approved):
Thank you for the unit tests :tada:
